### PR TITLE
exercise.js: remove jquery and convert to typescript

### PR DESCRIPTION
**Pull request by BramDevlaminck**
This pull request removes jquery code from exercise.js

A *feature* that is removed with this rewrite is the animation for the pop-up when a submission fails, since I did not find a way to replace `<div>.show("fast");`
The animation that is removed was this one:

https://user-images.githubusercontent.com/68779933/187682511-f1b42ad2-a0a9-4343-9e14-b0d38cbab63f.mov

progress on #3590  .

### CODE REVIEW

File: app/assets/javascripts/exercise.ts [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+function onFrameMessage(event): void {
+    if (event.message.type === "lightbox") {
+        showLightbox(event.message.content);
+    }
+}
+
+function initLightboxes(): void {
+    let index = 0;
+    const images = [];
+    document.querySelectorAll<HTMLElement>(".activity-description img, a.dodona-lightbox").forEach(el => {
```
Comment (niknetniko) [inline app/assets/javascripts/exercise.ts:29]:
```suggestion
    document.querySelectorAll<HTMLElement>(".activity-description img, a.dodona-lightbox").forEach((el, index) => {
```

The callback to the [`forEach`](https://developer.mozilla.org/en-US/docs/Web/API/NodeList/forEach) has an optional second argument, which is the index of the current element.

File: app/assets/javascripts/exercise.ts [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+                if (event.ctrlKey || event.metaKey || event.shiftKey || event.altKey) {
+                    return;
+                }
+                event.preventDefault();
+                loadFeedback(baseSubmissionsUrl + element.dataset.submission_id, element.dataset.submission_id);
+            });
+        });
+    }
+
+    function feedbackTableLoaded(userId: number, exerciseId: number, courseId: number): void {
```
Comment (niknetniko) [inline app/assets/javascripts/exercise.ts:270]: `courseId` can be undefined

File: app/assets/javascripts/types/index.d.ts [hunk 3]
```diff
@@ -16,4 +35,12 @@ declare module I18n {
...
 
-declare var dodona;
\ No newline at end of file
+// add parentIFrame from iFrame Resizer to the window to make typescript happy
+declare interface Window {
+    parentIFrame: any;
+}
+
+// ace editor is globally available
+declare var ace;
```
Comment (niknetniko) [inline app/assets/javascripts/types/index.d.ts:44]: I think that including the `@types/ace` package should take care of this (or at least, that's what it does in my editor, but unsure if they all do that; something to try).

**Review by niknetniko — CHANGES_REQUESTED** [review]

File: app/assets/javascripts/exercise.ts [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        window.MathJax.typeset([".gslide-description"]);
+    });
+    lightbox.open();
+
+    // Transfer focus back to the document body to allow the lightbox to be closed.
+    // https://github.com/dodona-edu/dodona/issues/1759.
+    document.body.focus();
+}
+
+function onFrameMessage(event): void {
```
[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/assets/javascripts/exercise.ts:20]: This is an event from the iFrameResizer library and there is no obvious way to define a good type for it. I could try to define a type that contains what I can see in the message send here but I don't know if this really adds a lot of extra information

[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/assets/javascripts/exercise.ts:20]: A solution is to use the `@types/iframe-resizer` library which has the appropriate types, so this problem is solved

[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/assets/javascripts/exercise.ts]: I've double-checked by logging the event in the console, and it's an `IFrameMessageData`-event from the library, but the message is custom (indeed from line 45).

The interface of the type from `@types/iframe-resizer`:
```ts
interface IFrameMessageData {
    iframe: IFrameComponent;
    message: any;
  }
```
You can see in the defined type that message can be whatever you want. In our case that is what is used on line 45.

File: app/assets/javascripts/exercise.ts [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+        });
+    }
+
+    function submitSolution(code: string): Promise<Response> {
+        const data = new FormData();
+        data.append("submission[code]", code);
+        data.append("submission[exercise_id]", exerciseId?.toString());
+        data.append("submission[course_id]", courseId?.toString());
+
+        return fetch("/submissions.json", {
```
[the remark this answers is not in the document]
Comment (BramDevlaminck) [reply app/assets/javascripts/exercise.ts:212]: ah this is exactly what I was looking for (I was already wondering if there was a function defined somewhere that did this), sweet!

**Review by niknetniko — APPROVED** [review]

**Review by jorg-vr — APPROVED** [review]

**Review by bmesuere — APPROVED** [review]

**Review by chvp — APPROVED** [review]

**Pull request merged**